### ISSUE TITLE
PayPal fails to work with Iframes

### ISSUE THREAD

**Issue by herbertmilhomme** [paypal/paypal-checkout#541]
Issue #140  Can we reopen it? I'm using a domain/url forward with masking, which loads page in an iframe. Not CORS, click-jacking or whatever. I want a sub-application to run off a secondary hostname without having to purchase a new webhost/server.

My issue is relevant to the original poster, as by default an iframe is created to render the page, and there's no logical way to reach top/parent-level of the webdesign. 

Need solutions... It's clear that this is an issue that's not resolved. Cause it was mentioned a few times already. But rather than fixing, it was closed out with a work-around hack.

I don't see why paypal shouldnt be compatible with iframes. Even if it was to prevent "click-jacking", the entire flow set-up is an overlay that takes over the page design, and a completely separate page is produced. Click-jacking or not, it should be obvious a transaction is being performed. Also, click-jacking is a security responsibility to the web-dev, not to a pay-pal plug-in. 

> If this whole issue is centered around security measures. Then it's a very disappointing mindset in my opinion for web-devs, to _control_ a user's experience. (Whether it's **forcing** password character restrictions, routinely pass resets, multi-layer auth procedures, etc... it's just not fun). As a developer using paypal's plug-in, i've had a lot of frustration, that i just resorted to the client-REST api, cause there either wasn't enough information/resources, or too many bugs with the others.

**Comment by bluepnume** [comment]
Thanks for the report. For what it's worth, the reason we require checkout.js is more UI related than for security reasons:

1. For cases when we render the experience inside an iframe, we generally need to render that iframe on the top-level page to ensure it has enough room to render

2. For the popup, we render an overlay over the top-level page to ensure users who de-focus the popup can't interact with the parent page, and have a way to re-focus the popup (by clicking on the overlay)

For both of these, checkout.js needs to be loaded on the top-level page so it can render there.

That said, I'm open to allowing rendering in an iframe if it meets our minimum dimensions. I'll keep this open as a feature request and try to find some time to prioritize this.

**Comment by herbertmilhomme** [comment]
In case you're wondering, the initial concern for the job i was doing was resolved by purchasing a webhost to bypass the hostname forwarding. Only cause i had a deadline and i was already past my time and mental energy with the project.

Future ideas and concepts still may follow similar design concept of being rendered on main site `www.example.com/sample/paypal-page` and displayed as `www.sample.com/paypal-page`.

But because a url masking/forwarding isnt the web-dev manipulating the files or content on the server. But more the way the DNS/Web-Host is directing to the source files, there isnt any real ways to add resources on that layer of the rendering process. Unless maybe some server code was the play a role in this. But i'm unfamiliar with any concept code that would render parent/top level when the page itself is what's being pulled in and streamed thru an iframe...

**Comment by bluepnume** [comment]
Hmm weird. At the very least I assume your server is rendering on the top-level page:

```html
<iframe src="..."></iframe>
```

How do you have this set up right now? Do you have a template with this iframe tag or something?

Ideally it would need to look like:

```html
<script src=".../checkout.js">
<iframe src="..."></iframe>
```

**Comment by herbertmilhomme** [comment]
Okay, i tried to take some screenshots to help to better explain the situation. 

![Url Masking](https://user-images.githubusercontent.com/15901670/33225902-aeb6b626-d14e-11e7-8034-e61bc7926bca.png) (A web-hosting service, not a web-dev skill)

![Url Masking Effects](https://user-images.githubusercontent.com/15901670/33225909-e1fe93e6-d14e-11e7-88fa-989449662b55.png)

It automatically forwards the url to a different hostname. This isnt web-code. It's server and protocalls and things a web-dev cant control with scripts. It's something that should be manipulated in PayPal to account for. Also, cant you just use a javascript `window` to check for full page width (in/outside of the iframe)?

**Comment by bluepnume** [comment]
I see. "Forward with masking" is automatically rendering that `<frame>` code and embedding the "forward to" domain.

The good news is, you could probably do the same manually -- it's not doing anything particularly special. You could create a static html page with the same html with a `<frame>` or an <iframe>, point it to the domain you want to embed, and add `checkout.js` in a script tag there.

(There are some disadvantages to embedding entire sites in iframes like this though. None of the urls for the embedded site will be visible in the url bar, for example.)

**Comment by herbertmilhomme** [comment]
I'd need to look into the server code to perform that action. I googled hostname forwarding and godaddy was the only thing I could locate. 

Any ideas for a asp.net/c# framework? (I'll accept php, tho I don't care much for it)

**Comment by bluepnume** [comment]
How are you currently creating / modifying your web pages? Are you writing any html by hand, or does your web host provide some kind of page editor that you're using?

**Comment by herbertmilhomme** [comment]
Writing by hand. I'm using .NET framework to code page in C#. I do understand what you mean about the doing it manually. I figured it could be done. But I couldn't locate any answers when I queried Google for results. Besides any pseudo code would be fine, just point me in the right direction.

**Comment by bluepnume** [comment]
So could you write a template in .NET to output the following html?

```html
<script src="https://www.paypalobjects.com/api/checkout.js"></script>
<iframe src="https://herbertmilhomme.com" style="width: 100%; height: 100%; border: 0"></iframe>
```

**Comment by herbertmilhomme** [comment]
Can you help me with one or the other?
**Problem:** Want's to forward domain to different hostname, but web-hosting server spawns the webpage inside an frame. 

**Solution 1:** Server side code to perform url forwarding that will also include the paypal js on the top layer (outside of frame)
**Response:** Okay, i understand _that_ part... but **how** do you forward a url hostname _and_ include the paypal? I tried to to google for source code to perform action and found no results. Do you have pseudo code on performing action using "server sided code" (i.e. c#, php, ruby, mvc, node.js, angular etc... not html)?

**Solution 2:** Paypal's js code should read client's window frame size in javascript/jquery, inside the `checkout.js` package?

>If it's _not_ possible with server-sided coding, _then_ logically the next option _is_ to manipulate paypal's code, right?

**Comment by bluepnume** [comment]
> Okay, i understand that part... but how do you forward a url hostname and include the paypal?

I'm a little confused at the question -- but you have two options:

1. Have the server send a `302` response with `Location: https://herbertmilhomme.com`. This redirects the browser to the new location, and since there are no iframes involved, `checkout.js` should work without any problems.

2. Embed the page in an iframe -- which is what you had previously set up in the godaddy control panel. You don't need any particular server-side code for this -- it's all achievable in plain html. You just need an html template containing both `<iframe src="https://herbertmilhomme.com"></iframe>` and `<script src="https://www.paypalobjects.com/api/checkout.js"></script>`. Any of the server-side languages you mentioned could render a simple html template with these two elements.

I think #2 is what you're going for right?

**Comment by herbertmilhomme** [comment]
So just to be on the same page, this can be done WITHOUT having to pay for an additional web-server? The original project have 1 server, split into 2 separate url domains. One is `.com` and the other is `.info`. When the user goes to `.info` they would be forwarded to a second folder, on the same server, but the url remains (`.info`). 

What you described there sounds exactly like the solution i performed that required me to spend MORE money than i intend to for this given task.

>In case you're wondering, the initial concern for the job i was doing was resolved by purchasing a _2nd_ webhost to bypass the hostname forwarding...

The whole reason for the url forwarding is to split/partition a single server into multiple url domains. Which isnt normally feasible without a dns server pointing to the correct path via ip address, which is also tied to a given directory path. The whole issue here is that i can not forward my hostname without the web-host displaying the webpage in an iframe. You said it's possible to do something server admins can do, from a web-developer stand-point.

**Note:** i am not admin of server (the computer that's housing the website), i'm merely the admin of the website itself (the person writing the code). If you're saying it's possible to perform this task, then you would need to include instructions on establishing and forwarding ip/hostname. Remember, this is something i googled, and led me to no results. :(

**Comment by herbertmilhomme** [comment]
Displaying a `301` redirect code isnt hard to do. But my _ONLY_ concern here is... If i'm throwing a ball (redirecting webpage) who's going to be on the other side to catch it? How does the server know to load `www.blah.com` upon redirect? I _need_ to assign the url to the path. That was what GoDaddy was doing for me.

**Comment by bluepnume** [comment]
Hi Herbert,

At a certain point there's only so much support I can give here for questions that go beyond rendering the PayPal button. It's your choice how you want to set up your redirects / domain forwarding. The only point I'm trying to make is, if you do want to go down the embedded-page-in-an-iframe path, it's entirely possible to:

1. Manually render the iframe in html (rather than using godaddy to do this automatically)
2. Place the `checkout.js` script alongside the iframe so the button can communicate with the parent page.

Otherwise it's your choice how you want to redirect / set up additional dns records to get your domains working the way you want.

If you see any specific problems with checkout.js after doing that, I'd be happy to help.

**Comment by herbertmilhomme** [comment]
Sorry. I was persistent on the idea cause I felt the direction you were trying to lead me down wasn't the best solution. I don't think that's an option that's available to me too pursue. Especially if it isn't a prominent result in frequent and multiple search queries, I just lack the confidence that it satisfies as a reasonable conclusion. 

If possible, I suggest we lean towards my alternate suggestion and create the PayPal checkout to rely on the actual window size. Regardless if inside an iframe or not, PayPal should

**Comment by bluepnume** [comment]
That's an option, but it's a longer term solution and it's a less optimal solution, since in that case the background overlay can't be displayed over the entire page.

If you're going with the godaddy "forward with masking" approach, you can achieve the exact same thing with a custom html template rendering an iframe and checkout.js script alongside each other. And if you're going with a 302 or new DNS rule, there's no iframe to cause a problem. So either way I think there's a path forward for you here.

**Comment by herbertmilhomme** [comment]
Just wanted to leave an update in response that i've ventured down the rabbit-hole and found that: 

The job of the DNS (Domain Name Server) is responsible for assigning an IP to Domain name relationship; it doesnt seem likely or possible to be able to alter and control such information _after_ the url has already been forwarded...

**Comment by JackTheHammer** [comment]
Hello, I am implementing PayPal on my service, and wanted to create following purchase flow:

1. When user clicks "pay with paypal" button, website opens a new browser window (e.g. window.open)
2. In this window, we have a hidden PayPal button (<form> tag) and an <iframe>. When page loads, form automatically posted to iframe and user sees this just like regular PayPal window.
3. Initial page (from step1) is visually shaded and have a links to bring this small js-created window to front or close it.

I've put PayPal page into iframe to have a better control over window, so I can not just track when it closed, but also move it around the screen. Otherwise, If I open paypal in it's own window, without iframe, I couldn't move it or chage it's size from code.

PayPal page loads just fine, but after user signs in, iframe redirects top-level window to PayPal and I loose full control over it's window (because of the browsers security policies)

If I set iframe's restrictions with sandbox attributes, the PayPal window doesn't login. I receive a message in browser console: 

> Uncaught DOMException: Failed to set the 'href' property on 'Location': The current window does not have permission to navigate the target frame to ...

And it points to following code block:
```
if (_integration.$integration.getConfig("REDIRECT_TOP") !== false || frameBust) {
    window.top.location = url;
 } else {
    window.location = url;
}
redirected = true;
```

Of course, I can just discard the issue and put PayPal in it's own window, withut iframe. But it seems that PayPal have a REDIRECT_TOP and frameBust flags, which I could use to prevent top-level window redirection.

The question is - how do I set this flags to false from top-level page and prevent it from being changed?
